sycl/include/sycl/ext/oneapi/annotated_arg/properties.hpp

```diff
@@ -327,15 +327,60 @@ struct is_valid_property<T, conduit_key::value_t> : std::true_type {};
 template <typename T>
 struct is_valid_property<T, stable_key::value_t> : std::true_type {};
 
+// Check if a given property is supported for annotated_arg
+// e.g. `alignment` is not supported for annotated_arg
+template <typename PropT>
+using is_property_value_of_annotated_arg =
+    is_property_value_of<PropT,
+                         annotated_arg<void, detail::empty_properties_t>>;
+
+// Check if a given property is supported for annotated_ptr
+template <typename PropT>
+using is_property_value_of_annotated_ptr =
+    is_property_value_of<PropT,
+                         annotated_ptr<void, detail::empty_properties_t>>;
+
+// Validate anntoated_arg/annotated_ptr
+template <typename T> struct validate_annotated_type : std::false_type {};
+
+// Partial specializations for validating annotated_arg
 template <typename T, typename... Props>
-struct check_property_list : std::true_type {};
+struct validate_annotated_type<annotated_arg<T, detail::properties_t<Props...>>>
+    : std::true_type {};
 
 template <typename T, typename Prop, typename... Props>
-struct check_property_list<T, Prop, Props...>
-    : std::conditional_t<is_valid_property<T, Prop>::value,
-                         check_property_list<T, Props...>, std::false_type> {
-  static_assert(is_valid_property<T, Prop>::value,
-                "Property is invalid for the given type.");
+struct validate_annotated_type<
+    annotated_arg<T, detail::properties_t<Prop, Props...>>>
+    : std::conditional_t<is_valid_property<T, Prop>::value &&
+                             is_property_value_of_annotated_arg<Prop>::value,
+                         validate_annotated_type<
+                             annotated_arg<T, detail::properties_t<Props...>>>,
+                         std::false_type> {
+  static_assert(
+      is_valid_property<T, Prop>::value,
+      "Property is invalid for the underlying type of annotated_arg.");
+  static_assert(is_property_value_of_annotated_arg<Prop>::value,
+                "Property is not supported for annotated_arg.");
+};
```

Comment:
Due to the `static_assert` inside the body, it seems unnecessary to check them in the conditional too. It _would_ make sense if `std::conditional` didn't evaluate both branches, but it does. However, I believe you can prevent further checks if you move the recursive use of `validate_annotated_type` to after the `static_assert`.

That is, I would prefer one of:
```suggestion
struct validate_annotated_type<
    annotated_arg<T, detail::properties_t<Prop, Props...>>>
    : validate_annotated_type<annotated_arg<T, detail::properties_t<Props...>>> {
  static_assert(
      is_valid_property<T, Prop>::value,
      "Property is invalid for the underlying type of annotated_arg.");
  static_assert(is_property_value_of_annotated_arg<Prop>::value,
                "Property is not supported for annotated_arg.");
};
```
or
```suggestion
struct validate_annotated_type<
    annotated_arg<T, detail::properties_t<Prop, Props...>>> {
  static_assert(
      is_valid_property<T, Prop>::value,
      "Property is invalid for the underlying type of annotated_arg.");
  static_assert(is_property_value_of_annotated_arg<Prop>::value,
                "Property is not supported for annotated_arg.");
  static constexpr bool value = validate_annotated_type<annotated_arg<T, detail::properties_t<Props...>>>::value;
};
```
where (to my knowledge) the latter would stop at the first invalid property.